### DISCUSSION TITLE
Question regarding the structure of the code (Role/Collection)

### DISCUSSION THREAD

**Discussion by robsonsnt** [autobase-tech/autobase#272] [General]
Hi,
First of all, I would like to thank you for making this code available to the community.
I would like to ask a question regarding the structure of the code. Have you considered making it available as a collection or a single role to make it possible to install with ansible-galaxy command, for example?
Is there any specific reason to work in the current form?
I am just trying to understand, thank you in advance!

**Comment by vitabaks** [comment]
Hi @robsonsnt 

thank you for your feedback

This project is published on the [galaxy](https://galaxy.ansible.com/vitabaks/postgresql_cluster) and you can install it

Example:

`ansible-galaxy install vitabaks.postgresql_cluster,1.10.0`

**Reply by robsonsnt** [reply]
Hi @vitabaks

thank you for the quick response
Let me try to explain in other words
I would like to ask why you didn't use the default Ansible collection structure.

**Reply by vitabaks** [reply]
I don't know why I need it. \
this project contains a lot of [roles](https://github.com/vitabaks/postgresql_cluster/tree/master/roles) (more than 30 roles)

Please suggest a structure that you see as more correct and describe what advantages it gives us.

**Reply by robsonsnt** [reply]
I understand, thank you for your contribution to the community, it really is a big project as mentioned with over 30 roles.
My intention was just to bring a constructive idea, in order to perhaps contribute with an idea for the project's evolution.
You have indeed been doing a great job contributing to the community.

**Reply by robsonsnt** [reply]
The "postgresql_cluster" repository is currently an Ansible role that installs and configures a PostgreSQL database cluster across multiple nodes. However, an Ansible collection can provide a more organized and scalable framework to deliver more complex functionality than just a role.

Here are some benefits the "postgresql_cluster" repository could gain by being transformed into an Ansible collection:

- Modularity: An Ansible collection allows various components such as roles, modules, and plugins to be organized into a single package. This makes it easier for users to manage and maintain their Ansible installations as they can use a single collection to include all necessary functionality.

- Compatibility: Ansible collections are designed to be compatible with other collections, making integration and use with other configuration management solutions easier. This also makes collaboration between different developers and teams on a single project easier.

- Reusability: An Ansible collection can be easily reused in different projects, saving time and effort in developing similar solutions. Individual roles can also be reused in other collections or projects, allowing developers to leverage the expertise of other projects and avoid reinventing the wheel.

- Flexibility: An Ansible collection can include multiple components such as roles, modules, and plugins, allowing users to customize and extend functionality. This enables users to adapt collections to their specific needs.

In summary, transforming the "postgresql_cluster" repository into an Ansible collection can provide significant benefits in terms of organization, compatibility, reusability, and flexibility. However, transforming it into a collection will require significant development effort to refactor the existing role into a complete collection package.

**Reply by vitabaks** [reply]
At the moment, a few more sponsors are needed so that I can devote more time to improving this project.  
Thanks for your comments, I've taken that into account.

**Reply by SamuelMolling** [reply]
I can try to help with that.

**Reply by vitabaks** [reply]
Thank you for responding.  

The fact is that I still do not quite understand why the current structure with multiple roles may not be convenient.  Perhaps I need to learn more about collections, this is a fairly new term in ansible.

**Reply by ThomasSanson** [reply]
Hi @vitabaks,

I believe that migrating to the "collection" format in Ansible presents several advantages for the PostgreSQL High-Availability Cluster project. Here are some of the main advantages:

1. **Better organization and modularity**: Collections allow you to group multiple Ansible roles, plugins, and other components into a single entity. This makes it easier to manage and maintain the project.

2. **Reuse and sharing**: Collections can be easily shared and reused within the Ansible community, which helps increase the visibility and adoption of the project.

3. **Dependency management**: With collections, it's easier to manage dependencies between different components of the project, which simplifies installation and updates.

4. **Better integration with Ansible Galaxy**: Collections are designed to work closely with Ansible Galaxy, which allows for better integration with the platform and simplifies the installation process for users.

An example of a project that has made this transformation is the Confluent Platform Ansible project: https://github.com/confluentinc/cp-ansible. You can see how they have organized their roles, plugins, and other components into a single collection, which greatly facilitates the management and use of the project.

I understand that the concept of collections may be new to some Ansible users, and it's completely normal to wonder why to change the current project structure.

Although the current structure based on multiple roles works, collections offer a more structured and modular organization. Collections help to group and manage Ansible roles, plugins, and other components into a single entity.

Using collections makes it easier to share and reuse the project within the Ansible community. Additionally, collections offer better integration with Ansible Galaxy, which simplifies installation and updates for users.

It's true that migrating to collections may require some learning and adaptation time, but it presents long-term advantages for the management and maintenance of the project.

I encourage you to learn more about collections and to consider this migration to improve the user experience and maintainability of the PostgreSQL High-Availability Cluster project.

I hope this gives you a better idea of the advantages of migrating to the "collection" format for the PostgreSQL High-Availability Cluster project. Thank you for considering this request.

Best regards,
ThomasSanson

**Reply by ThomasSanson** [reply]
@vitabaks,

I would like to add some additional points to highlight the benefits of transitioning to the Ansible collection format for the PostgreSQL High-Availability Cluster project:

- Transitioning to a collection will allow this project to be fully referenced under Ansible Galaxy. This will make the project more discoverable, and make installation and updates easier for users.
- Users will no longer need to clone or modify the repository, as they can directly use the playbooks and roles from the collection. This simplifies and accelerates the integration of the project into existing environments.
- Transitioning to the collection format can make it easier for enterprise users to contribute. Enterprises are often more willing to allocate official work time to contribute to projects structured as collections.
- Migrating to collections also facilitates the introduction of better testing for the playbook. This will enable the inclusion of installation, recovery, and testing playbooks.
- I encourage you to look at the Confluent Platform (Kafka) project, which has also adopted the collection format and why they made this choice. Their experience can serve as an example for the migration of this project.

I am aware that transitioning to the collection format requires a lot of work, but I am ready to take on this task to ease the workload on your end and ensure a smooth transition.

Best regards,
ThomasSanson

**Reply by vitabaks** [reply]
> I am aware that transitioning to the collection format requires a lot of work, but I am ready to take on this task to ease the workload on your end and ensure a smooth transition.

Thanks @ThomasSanson for the detailed description! And for the fact that you want to contribute to the project. 

For me, the main goal has always been to simplify the use of this automation as much as possible for the user, even for someone who is not familiar with ansible at all. That's why I described a detailed README and all the [steps](https://github.com/vitabaks/postgresql_cluster#deployment-quick-start) for implementing a cluster from scratch.  

I think you see more advantages in using the collections format than I do because (apparently) that you understand this better. Let's try to do it.

But the main goal - everything should be as simple and clear as possible for users.

**Reply by pat-s** [reply]
The collection seems out-of-date (2.0.0-rc1) vs 2.1.0 as the latest release. It is planned to updated and curated in the future?

(the use of a collection would make using different environments (as mentioned in #869 a lot easier)

**Reply by vitabaks** [reply]
> The collection seems out-of-date (2.0.0-rc1) vs 2.1.0 as the latest release. It is planned to updated and curated in the future?
> 
> (the use of a collection would make using different environments (as mentioned in #869 a lot easier)

Autobase is not just an Ansible collection; it is a full-fledged DBaaS platform that leverages Ansible as one of its components for database automation. Because of this, we do not aim for full compatibility with the proposed format.

Regarding support and future updates, development is ongoing—feel free to check the latest commits in the master branch.

**Reply by pat-s** [reply]
I understood that, but that doesn't answer the question.

As autobase offers and promotes the CLI-ansible installation approach and a collection is available (which is great!), it would be great if this one could be updated and supported as releases are issued.

I'd prefer to use the collection over a submodule checkout (my current approach) and for simplified multi-environment usage and integration into a broader ansible approach.

**Reply by vitabaks** [reply]
> it would be great if this one could be updated and supported as releases are issued

Perhaps I don’t understand you, because the code is already supported and we public new releases (about once every 3 months)

**Reply by pat-s** [reply]
https://galaxy.ansible.com/ui/standalone/roles/vitabaks/postgresql_cluster/ is out of date and in general it would be great to have a collection which contains all roles in https://github.com/vitabaks/autobase/tree/master/automation. This would lean more towards a standard ansible flow. Checking out a git repo and calling included playbooks that change over time is quite uncommon in the ansible world (NB: I am saying this managing multiple Ansible Collections on Galaxy).

Currently, playbooks are large and more like a role itself. The best structure would likely be to migrate the playbooks (like `config_pgcluster.yml`) to roles. After installation the collection then, users could call a role from the collection with variables defined in `group_vars` or `host_vars`. This would be a lot more transparent. Example:

```yaml
- name: Postgres
  hosts: all
  become: true
  gather_facts: true

  roles:
    - role: autobase.deploy_pgcluster
    - role: autobase.configure_pgcluster
    [...]
```

In addition, this allow defining `autobase` as a proper dependency in `requirements.yml` and allow for automated versioning and updates of dependency versioning tools.

**Reply by vitabaks** [reply]
I understand your point. Indeed, the `postgresql_cluster` on Galaxy is no longer updated because it has evolved into `Autobase`, which we now position as a DBaaS solution with UI-based management.

I see the value in aligning more closely with the standard Ansible flow, but I’m not sure how much this would contribute to the project’s growth or adoption. Would it lead to wider adoption and increased sales?

If this is something you find important, I encourage you to create a [feature request](https://github.com/vitabaks/autobase/issues), and we’ll evaluate it. However, I can’t promise a high priority for this at the moment, as there are other critical features in development. That being said, if a [sponsor](https://autobase.tech/docs/sponsor) is willing to support this effort, we would be happy to consider it sooner.

**Reply by pat-s** [reply]
> I see the value in aligning more closely with the standard Ansible flow, but I’m not sure how much this would contribute to the project’s growth or adoption. Would it lead to wider adoption and increased sales?

If `autobase` would be become more easy and straightforward to use (which includes simplified embedding into existing Ansible structures), I would expect this, yes.

> If this is something you find important, I encourage you to create a [feature request](https://github.com/vitabaks/autobase/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen), and we’ll evaluate it. However, I can’t promise a high priority for this at the moment, as there are other critical features in development. That being said, if a [sponsor](https://autobase.tech/docs/sponsor) is willing to support this effort, we would be happy to consider it sooner.

I guess the overarching question is whether this is a feature or actually a bug, i.e. whether it should have been like this from the start (in an ideal scenario). (Also, subconsciously, the "please donate then we do work" banner is being placed a bit too often in some discussions I've visited here recently).

Using Ansible directly is surely on the expert side and not targeted towards users preferring the UI unless it provides *substantial* benefits. Maybe `autobase` needs to make a decision where it wants to go or what it wants to be? (also in terms of open/closed support)
If `autobase` should be usable directly through Ansible in the longterm as a "goodie" alongside users preferring the UI (and possibly more often requesting help), this would be great. Eventually, these users will also start donating then.

Take me as an example user/client: I am used to work with Ansible and do not need a/the UI. What I need is a thorough and robust way to configure and maintain a (HA)-DB deployment(s). I am willing to donate/pay for it if there is a solid and professional way to do so. Right now I am playing around with `autobase` and I am unsure if the current approach (deployment-wise) is a good one which I can/want to go with longterm. 

Aside from this, I also see advantages for a clear collection/role structure for the project itself. In the end, Ansible is the base and it also gets applied somehow in the background. Doing so by following best practices should also simplify the internal use.

I don't know your Ansible skills. Yet, given you opted for it as the core tool of this project, they should be quite solid. Hence, I hope you understand my proposal (it is only meant to be positive and helpful). In case you need help with a potential restructuring, I am also happy to help.

**Reply by vitabaks** [reply]
Feel free to submit a PR if you believe it will improve the project without introducing additional inconvenience for long-time users who have been using the current version for years.

Additionally, any changes should not disrupt the existing [Console API](https://github.com/vitabaks/autobase/tree/master/console/service) logic, which executes playbooks rather than roles.

**Reply by pat-s** [reply]
It will surely be a breaking change in terms of usage and versioning. The important bit though is that you must be convinced about the change, otherwise it doesn't have a base to be operated on in the future.
Ideally the change would be limited to calling a role instead of a playbook and placing the vars differently. 

In addition, I would suggest to use a prefix for roles which are user facing to differentiate them from the ones which are internal (if they should be kept as roles like now). (Ideally, there would only be user-facing roles but achieving this would be an even bigger change.)

> Additionally, any changes should not disrupt the existing [Console API](https://github.com/vitabaks/autobase/tree/master/console/service) logic, which executes playbooks rather than roles.

It can still call a playbook but the playbook itself then simply needs to execute the roles instead of the tasks directly.

**Comment by vitabaks** [comment]
While we use Ansible as one of the components in our solution, **Autobase is more than just an Ansible role or collection—it is a DBaaS solution**. As such, we are not strictly bound by standard Ansible structures, though we follow best practices where they align with our goals.  

Our primary focus is on delivering robust database automation and a seamless user experience, which sometimes requires deviations from conventional Ansible patterns.
